charts/dapr/crds/configuration.yaml

```diff
@@ -375,11 +316,7 @@ spec:
                     properties:
                       endpointAddress:
                         type: string
-                    required:
```

Comment:
I was able to make most fields appear as required again. These are fields that are placed in arrays or dictionaries, so it's ok if they don't have "omitempty" as it shouldn't impact the functionality I was concerned with.

Reply:
> This should be fine however, as we never relied on kubernetes native validation much

Literally every user and system using Dapr CRDs rely on K8s native validation 100% of the time. Please return the required fields. This has nothing to do with the struct definitions.

Reply:
All the fields that could be made required are there. I am afraid I can't do more than this.

Controller-gen assumes a field is not required if it has the "omitempty" tag, and anything I tried to override that didn't work.

Reply:
Ok, to be clear - if there are any fields in the spec that were required and enforced before this change and are now not required and were made optional, that's a breaking change to the spec and a regression. If that's not the case, we can go forward with this change.

Reply:
Making a field that was optional, required is a breaking change, because things that worked before may not anymore. Changing fields from required to optional is not a breaking change, because what worked before will continue to work.

In this case, I don't think there's really any net effect, because before we were not requiring these fields anyways.

A good example is that the field `mtls.enabled` used to be required and now it's not anymore: https://github.com/dapr/dapr/pull/6101/files#diff-98b7452caa61c9b7f101305e14dc06249d70a8a2fe93b806857e1317ac1a68bdL314-L324 However, `mtls` itself was not required (or users would have had to specify a `mtls` field in every Configuration resource!): what that meant is that before, if users wrote `mtls` then they **had** to also type `mtls.enabled` explicitly. This should not cause any issue to users, and could actually make things easier if users try to combine multiple Configuration resources in which one enables mTLS and the other one sets the clock skew, for example.

All other fields that have the "required" tag removed are in the same boat.

Reply:
>  Changing fields from required to optional is not a breaking change, because what worked before will continue to work.

Disagree, in that case invalid YAML definitions that would lead to a misconfigured state that would have been previously caught at the API level now become runtime level errors, so that's a definite regression to user experience that we cannot allow.

> because before we were not requiring these fields anyways.

This is what matters - if these fields were marked as required but not treated as such before, then this change is ok. I'll use the modified CRDs from this PR and test it out to validate.

Reply:
When testing the modified CRDs it's clearly observed that there are breaking changes to the specs of all CRDs. As one example, with the modified CRDs it is possible to deploy the following YAML which omits the `endpointAddress` field, whereas in 1.11 this would result in an error.

```yaml
apiVersion: dapr.io/v1alpha1
kind: Configuration
metadata:
  name: tracing
spec:
  tracing:
    samplingRate: "1"
    otel:
      isSecure: false
      protocol: grpc
```

Reply:
> with the modified CRDs it is possible to deploy the following YAML which omits the endpointAddress field, whereas in 1.11 this would result in an error.

How is that a breaking change though? If it were the other way around (before it could be deployed, now it can't), it would be, but this is making something more lax.

In this case, otel would not be enabled because the endpoint is missing. We have never relied on K8s validation alone because it doesn't work in standalone mode.

Reply:
It's a breaking change to the CRD spec as it can and will result in any type of validation to now allow misconfigurations into the cluster state where it would once break.

To focus on the actual issue and avoid discussion over semantics, even if for the sake of the argument we say there's no breaking change here, this is just as extremely unacceptable as it results in a degraded user experience that leads users down the non-happy path. Kubernetes users rely on them to prevent misconfiguration using native tooling. Required fields are an extremely common thing in schemas with a clear purpose of explicitly describing what needs to be filled to arrive at a working desired state.

Reply:
How would you recommend we make this thing work then?